File: solutions/python/matching-brackets/1/matching_brackets.py
def is_paired(input_string):
    brackets = []

    mapping = {
        '(': ')',
        '[': ']',
        '{': '}',
    }
    
    for char in input_string:
        if char in ['[', '{', '(']:
            brackets.append(char)
        elif char in [']', '}', ')']:
            if not brackets:
                return False

            last_open = brackets.pop()
            if char != mapping[last_open]:
                return False

    return brackets == []

File: solutions/vbnet/matching-brackets/1/MatchingBrackets.vb
Imports System.Collections.Generic

Public Module MatchingBrackets
    Public Function IsPaired(ByVal input As String) As Boolean
        IsPaired = true

        Dim brackets As New Stack(Of Char)()

        For Each cc As Char In input
            Select Case cc
                Case "{"c
                    brackets.Push("}"c)
                Case "["c
                    brackets.Push("]"c)
                Case "("c
                    brackets.Push(")"c)
                Case "}"c, "]"c, ")"c
                    If brackets.Count = 0
                        IsPaired = false
                    Else
                        Dim prev as Char
                        prev = brackets.Pop()
                        If prev <> cc 
                            IsPaired = false
                        End If
                    End If
            End Select
        Next

        If brackets.Count = 0
            IsPaired = IsPaired and true
        Else
            IsPaired = false
        End If

    End Function
End Module
